### PR TITLE
Cmdlets for managed service identity on Application Gateway

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Description
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/220

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 [hunk 1]
```diff
@@ -14,25 +14,25 @@
 
 function Test-AvailableSslOptions
 {
-	$result = Get-AzApplicationGatewayAvailableSslOptions
```
Comment (cormacpayne) [inline src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: @akshaysngupta any reason the prefixes were changed from `Az` back to `AzureRM`?

File: src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 [hunk 2]
```diff
@@ -784,97 +785,122 @@ function Test-ApplicationGatewayCRUD3
...
-		$vnet = New-AzvirtualNetwork -Name $vnetName -ResourceGroupName $rgname -Location $location -AddressPrefix 10.0.0.0/16 -Subnet $gwSubnet
-		$vnet = Get-AzvirtualNetwork -Name $vnetName -ResourceGroupName $rgname
-		$gwSubnet = Get-AzVirtualNetworkSubnetConfig -Name $gwSubnetName -VirtualNetwork $vnet
+		$gwSubnet = New-AzureRmVirtualNetworkSubnetConfig -Name $gwSubnetName -AddressPrefix 10.0.0.0/24
+		$vnet = New-AzureRmvirtualNetwork -Name $vnetName -ResourceGroupName $rgname -Location $location -AddressPrefix 10.0.0.0/16 -Subnet $gwSubnet
+		$vnet = Get-AzureRmvirtualNetwork -Name $vnetName -ResourceGroupName $rgname
+		$gwSubnet = Get-AzureRmVirtualNetworkSubnetConfig -Name $gwSubnetName -VirtualNetwork $vnet
+
+		# Create Managed Identity
+		$identity = New-AzureRmUserAssignedIdentity -Name $identityName -Location $location -ResourceGroup $rgname
```
Comment (cormacpayne) [inline src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: @akshaysngupta if you're using the MSI cmdlets in your tests, you'll want to make sure that you add the MSI project to the Network solution so that it gets built along with Network

File: src/Network/Network/ChangeLog.md [hunk 3]
```diff
@@ -21,9 +21,16 @@
...
+    - New cmdlets added:
+        - Set-AzApplicationGatewayIdentity
+        - Get-AzApplicationGatewayIdentity
+        - New-AzApplicationGatewayIdentity
+        - Remove-AzApplicationGatewayIdentity
+    - New-AzApplicationGateway cmdlet updated with optional parameter -Identity
+>>>>>>> change log
 
 ## Version 1.0.0
-* General availability of `Az.Network` module
```
Comment (cormacpayne) [inline src/Network/Network/ChangeLog.md]: @akshaysngupta please revert the change made to this line

File: src/Network/Network/ChangeLog.md [hunk 4]
```diff
@@ -21,9 +21,16 @@
 
 ## Version 1.1.0
 * Update incorrect online help URLs
+* Added Cmdlets for Identity on Application Gateway.
```
Comment (cormacpayne) [inline src/Network/Network/ChangeLog.md]: @akshaysngupta the content you added here should be under the `Upcoming Release` header

File: src/Network/Network/ChangeLog.md [hunk 5]
```diff
@@ -21,9 +21,16 @@
...
+    - New cmdlets added:
+        - Set-AzApplicationGatewayIdentity
+        - Get-AzApplicationGatewayIdentity
+        - New-AzApplicationGatewayIdentity
+        - Remove-AzApplicationGatewayIdentity
+    - New-AzApplicationGateway cmdlet updated with optional parameter -Identity
+>>>>>>> change log
```
Comment (cormacpayne) [inline src/Network/Network/ChangeLog.md]: @akshaysngupta please remove this line

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 6]
```diff
@@ -190,6 +190,13 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
         [Alias("UserAssignedIdentity")]
         public string UserAssignedIdentityId { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Application Gateway Identity to be assigned to Application Gateway.")]
+        [ValidateNotNullOrEmpty]
+        public PSManagedServiceIdentity Identity { get; set; }
```
Comment (cormacpayne) [inline src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs]: @akshaysngupta it looks like there exists a `-UserAssignedIdentityId` parameter in this cmdlet that does exactly what `New-AzApplicationGatewayIdentity` does, and they aren't in separate parameter sets, so users will be able to provide a value for both. Does it make sense to have this parameter if it does the same thing as the new cmdlet you're adding, but doesn't require an additional cmdlet call?

File: src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 (hunk 1 above)
Comment (akshaysngupta) [reply src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: This wasn't intended. I have reverted it back.

File: src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 (hunk 2 above)
Comment (akshaysngupta) [reply src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: ManagedServiceIdentity proj is already in Network.sln

File: src/Network/Network/ChangeLog.md (hunk 3 above)
Comment (akshaysngupta) [reply src/Network/Network/ChangeLog.md]: corrected.

File: src/Network/Network/ChangeLog.md (hunk 4 above)
Comment (akshaysngupta) [reply src/Network/Network/ChangeLog.md]: corrected.

File: src/Network/Network/ChangeLog.md (hunk 5 above)
Comment (akshaysngupta) [reply src/Network/Network/ChangeLog.md]: removed.

File: src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 (hunk 1 above)
Comment (cormacpayne) [reply src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: @akshaysngupta no new commit was pushed to your branch, so this is still an issue

File: src/Network/Network/ChangeLog.md (hunk 3 above)
Comment (cormacpayne) [reply src/Network/Network/ChangeLog.md]: @akshaysngupta same comment -- no new commit, so this is still an issue

File: src/Network/Network/ChangeLog.md (hunk 4 above)
Comment (cormacpayne) [reply src/Network/Network/ChangeLog.md]: @akshaysngupta same comment -- no new commit, so this is still an issue

File: src/Network/Network/ChangeLog.md (hunk 5 above)
Comment (cormacpayne) [reply src/Network/Network/ChangeLog.md]: @akshaysngupta same comment -- no new commit, so this is still an issue

**Comment by vladimir-shcherbakov** [comment]
@akshaysngupta Ping on the comments above.

File: src/Network/Network/help/New-AzApplicationGateway.md [hunk 7]
```diff
@@ -30,7 +30,8 @@ New-AzApplicationGateway -Name <String> -ResourceGroupName <String> -Location <S
  [-RedirectConfigurations <PSApplicationGatewayRedirectConfiguration[]>]
  [-WebApplicationFirewallConfiguration <PSApplicationGatewayWebApplicationFirewallConfiguration>]
  [-AutoscaleConfiguration <PSApplicationGatewayAutoscaleConfiguration>] [-EnableHttp2] [-EnableFIPS]
- [-Zone <String[]>] [-Tag <Hashtable>] [-UserAssignedIdentityId <String>] [-Force] [-AsJob]
+ [-Zone <String[]>] [-Tag <Hashtable>] [-UserAssignedIdentityId <String>]
```
Comment (cormacpayne) [inline src/Network/Network/help/New-AzApplicationGateway.md]: @akshaysngupta you'll need to regenerate the help for this cmdlet

File: src/Network/Network/help/New-AzApplicationGatewayIdentity.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+## DESCRIPTION
+**New-AzApplicationGatewayIdentity** cmdlet creates an application gateway identity object.
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\>$identity = New-AzUserAssignedIdentity -Name $identityName -ResourceGroupName $rgName -Location $location
+PS C:\>$appgwIdentity = New-AzApplicationGatewayIdentity -UserAssignedIdentity $identity.Id
```
Comment (cormacpayne) [inline src/Network/Network/help/New-AzApplicationGatewayIdentity.md]: @akshaysngupta this example should be updated to use the result of this cmdlet in `New-AzApplicationGateway` to show the end-to-end flow

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 9]
```diff
@@ -183,13 +183,22 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
         public Hashtable Tag { get; set; }
 
         [Parameter(
+            ParameterSetName = "IdentityByUserAssignedIdentityId",
             Mandatory = false,
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "ResourceId of the user assigned identity to be assigned to Application Gateway.")]
         [ValidateNotNullOrEmpty]
         [Alias("UserAssignedIdentity")]
         public string UserAssignedIdentityId { get; set; }
```
Comment (cormacpayne) [inline src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs]: @akshaysngupta since the intention is to remove this cmdlet, you will want to add our [custom breaking change attribute](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#a-parameter-is-being-deprecated) on this parameter so users are notified of the change whenever they use it.

You should also add an entry to the [`upcoming-breaking-changes.md`](https://github.com/Azure/azure-powershell/blob/master/src/Network/documentation/upcoming-breaking-changes.md) file in the Network folder to call out this change (you can remove the content already in the file)

File: src/Network/Network/ApplicationGateway/Identity/RemoveAzureApplicationGatewayIdentityCommand.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        [Parameter(
+             Mandatory = true,
+             ValueFromPipeline = true,
+             HelpMessage = "The applicationGateway")]
+        public PSApplicationGateway ApplicationGateway { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Do not ask for confirmation.")]
+        public SwitchParameter Force { get; set; }
```
Comment (cormacpayne) [inline src/Network/Network/ApplicationGateway/Identity/RemoveAzureApplicationGatewayIdentityCommand.cs]: @akshaysngupta as discussed offline -- please remove this parameter and regenerate the help

**Comment by cormacpayne** [comment]
Waiting for the [release branch to be merged back into `master`](https://github.com/Azure/azure-powershell/pull/8510) before merging this PR
